umbra/umbra_32.c

```diff
@@ -476,6 +476,7 @@ shadow_table_init(umbra_map_t *map)
                                           HEAPSTAT_SHADOW);
     } else {
         map->shadow_table = static_shadow_table;
+        static_shadow_table_unused = true;
```

Comment:
So the idea is to use the static table for the first mapping, and if someone else wants a new mapping, to dynamically allocate.  Could we rename `s/static_shadow_table_unused/static_shadow_table_claimed/` or sthg to make this clearer and hopefully avoid things like https://github.com/DynamoRIO/drmemory/pull/2044#discussion_r140582790 where it is mis-interpreted.

To be safe we may want a mutex here as well, or else to document that this can only be called at process init with a single thread (i.e., document it's not thread-safe).

Alternatively, zhaoqin@ suggests removing the static table and the flag altogether and always using dynamic alloc, which seems fine to me as well.

Reply:
I'm personally not a big fan of global variables and use them only when necessary. I like @zhaoqin's suggestion. It makes the code cleaner with less branching.